### PR TITLE
Basic OJO analysis

### PR DESCRIPTION
**Pull request by lizgzil**
## OJO analysis

All the ojo analysis and the associated scripts are now in the folder `ojd_daps_skills/analysis/OJO/`

There are various scripts to process the ojo output into the formats we'll need for analysis (see the [README](https://github.com/nestauk/ojd_daps_skills/blob/86548bed6d42817a8e2db34f318907dd50d3c6f3/ojd_daps_skills/analysis/OJO/README.md)).

Main thing for review: 
- Check whether the `ojd_daps_skills/analysis/OJO/process_analysis_data.py` script is ok
This script creates the two files neccessary for (hopefully) all of our analysis:
1. The file s3://open-jobs-lake/escoe_extension/outputs/data/model_application_data/dedupe_analysis_metadata.csv
Contains collated metadata for all 3,693,313 deduplicated job adverts (with a 6 week stock)
2. The file s3://open-jobs-lake/escoe_extension/outputs/data/model_application_data/dedupe_analysis_skills_sample.json (and the .csv - which is the same data just in a different file format) contains a random sample of the metadata + skills data for 100,000 job adverts from the 3,693,313 deduplicated job adverts.

Then the high level analysis is done in this notebook - https://github.com/nestauk/ojd_daps_skills/blob/86548bed6d42817a8e2db34f318907dd50d3c6f3/ojd_daps_skills/analysis/OJO/ojo_data_summary.ipynb

@india-kerle you'll see I removed the bits we had to format the skills when they were in that weird format.

---

Checklist:

- [ ] I have refactored my code out from `notebooks/`
- [ ] I have checked the code runs
- [ ] I have tested the code
- [ ] I have run `pre-commit` and addressed any issues not automatically fixed
- [ ] I have merged any new changes from `dev`
- [ ] I have documented the code
  - [ ] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [ ] I have explained the feature in this PR or (better) in `output/reports/`
- [ ] I have requested a code review

### CODE REVIEW

File: ojd_daps_skills/utils/filter_outputs.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
+"""
+One off script to filter the output results for just the sample of job adverts.
+The saved out files will be easier to load for any analysis.
```
Comment (lizgzil) [inline ojd_daps_skills/utils/filter_outputs.py]: fyi @india-kerle

File: ojd_daps_skills/utils/filter_outputs.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+save_to_s3(
+    s3,
+    bucket_name,
+    duplicates_sample,
+    "escoe_extension/outputs/data/model_application_data/job_ad_duplicates_sample.csv",
+)
+
+skills_file_name = "ojd_daps_skills/utils/job_ad_to_skills.json"  # Needed to be downloaded from S3 and saved locally since its so big
+skills_data = load_file(skills_file_name, s3=False)
+skills_data_sample = [s for s in skills_data if int(s["job_id"]) in raw_job_adverts_ids]
```
Comment (lizgzil) [inline ojd_daps_skills/utils/filter_outputs.py]: @india-kerle after loading, this is all I did to filter so shouldn't have done anything weird!

File: ojd_daps_skills/utils/filter_outputs.py (hunk 1 above)
Comment (india-kerle) [reply ojd_daps_skills/utils/filter_outputs.py]: i think we can add the cleaning of the output to this script as well i.e. getting the first element of the list and converting it to a tuple

File: ojd_daps_skills/utils/duplication.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        while curr_date < end:
+            date_chunk_mapper[curr_date.strftime("%Y-%m-%d")] = start.strftime(
+                "%Y-%m-%d"
+            )
+            curr_date += timedelta(days=1)
+        date_chunk_mapper[curr_date.strftime("%Y-%m-%d")] = start.strftime("%Y-%m-%d")
+    return date_chunk_mapper
+
+
+def get_deduplicated_job_adverts(
```
Comment (lizgzil) [inline ojd_daps_skills/utils/duplication.py]: @Jack-Vines this is how I've dealt with duplication (using s3://open-jobs-lake/escoe_extension/outputs/data/model_application_data/raw_job_adverts_sample.csv and s3://open-jobs-lake/escoe_extension/outputs/data/model_application_data/job_ad_duplicates_sample.csv) 
Please could you check it over and I'm happy to chat if anything is unclear?

FYI @india-kerle

File: ojd_daps_skills/utils/plotting.py [hunk 4]
```diff
@@ -8,10 +8,6 @@
 
 ChartType = alt.vegalite.v4.api.Chart
 
-# Fonts and colours
-FONT = "Averta"  # should be changed depending on what your font is called in your system's font book
```
Comment (india-kerle) [inline ojd_daps_skills/utils/plotting.py:12]: can you believe I found out how everyone got averta font saved to their fontbook - there's a shared folder called 'Nesta new Visual Identity Sept 2021' that has the Averta font that saves to 'Averta' - who knew...!

Comment (india-kerle) [reply ojd_daps_skills/utils/plotting.py:12]: although I still think we should pass this as a parameter as people can save the name differently i.e. like you and i did

File: ojd_daps_skills/analysis/OJO/process_analysis_data.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+        merged_metadata,
+        os.path.join(args.s3_folder, args.metadata_output),
+    )
+
+    # Sample skills data
+    sample_n = 100000
+    # Shuffle, then take the first 100000 which are in the dedupe list
+    job_skills_shuffled = job_skills.copy()
+    random.seed(42)
+    random.shuffle(job_skills_shuffled)
```
Comment (india-kerle) [inline ojd_daps_skills/analysis/OJO/process_analysis_data.py:145]: this is a good idea

File: ojd_daps_skills/analysis/OJO/process_analysis_data.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
+"""
+Filter job data using deduplicated job ids to create outputs for the analysis pieces
+Output:
```
Comment (india-kerle) [inline ojd_daps_skills/analysis/OJO/process_analysis_data.py]:
```suggestion
Outputs:
```

File: requirements.txt [hunk 7]
```diff
@@ -20,4 +20,5 @@ selenium==4.2.0
 altair_saver
 webdriver-manager
 s3fs
-https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-3.1.0/en_core_web_sm-3.1.0.tar.gz
\ No newline at end of file
+https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-3.1.0/en_core_web_sm-3.1.0.tar.gz
```
Comment (india-kerle) [inline requirements.txt:23]: think this will be a separate PR but we should triple check all spaCy requirements given @Jack-Vines 's doozy of a time with it

File: ojd_daps_skills/analysis/OJO/README.md [hunk 8]
```diff
@@ -0,0 +1,48 @@
+# OJO analysis
+
+The analysis of skills data was done by applying the skills extraction class to a large sample of job adverts in Nesta's Open Jobs Observatory (OJO). Skills were extracted from these adverts and the data was saved out. The data from OJO is not available in the public S3 bucket, and thus this analysis is only for internal Nesta purposes.
+
+## Deduplication of job adverts
+
+To get the ids of deduplicated job adverts run the `ojd_daps_skills/analysis/deduplicate_job_ids.py` script with correct arguments.
+
+For the full dataset and a 6 week time chunk for job stocks (this takes about 4 hours to run):
```
Comment (india-kerle) [inline ojd_daps_skills/analysis/OJO/README.md:9]: yeowza :0

**Review by india-kerle — APPROVED** [review]
no major red flags! Can't believe it took so long to deduplicate and merge all the metadata :0 - thanks so much for doing that. Looks like this could be really helpful for our up coming poster.
